src/main/java/at/ac/tuwien/kr/alpha/common/program/impl/InternalProgram.java

```diff
@@ -0,0 +1,103 @@
+package at.ac.tuwien.kr.alpha.common.program.impl;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.commons.lang3.tuple.ImmutablePair;
+
+import at.ac.tuwien.kr.alpha.common.Predicate;
+import at.ac.tuwien.kr.alpha.common.atoms.Atom;
+import at.ac.tuwien.kr.alpha.common.program.AbstractProgram;
+import at.ac.tuwien.kr.alpha.common.rule.impl.InternalRule;
+import at.ac.tuwien.kr.alpha.common.rule.impl.NormalRule;
+import at.ac.tuwien.kr.alpha.grounder.FactIntervalEvaluator;
+import at.ac.tuwien.kr.alpha.grounder.Instance;
+
+/**
+ * A program in the internal representation needed for grounder and solver, i.e.: rules must have normal heads, all aggregates must be rewritten, all intervals
+ * must be preprocessed (into interval atoms), equality predicates must be rewritten
+ * 
+ * Copyright (c) 2017-2019, the Alpha Team.
+ */
+public class InternalProgram extends AbstractProgram<InternalRule> {
+
+	private final Map<Predicate, HashSet<InternalRule>> predicateDefiningRules = new LinkedHashMap<>();
+	private final Map<Predicate, LinkedHashSet<Instance>> factsByPredicate = new LinkedHashMap<>();
+	private final Map<Integer, InternalRule> rulesById = new LinkedHashMap<>();
+
+	public InternalProgram(List<InternalRule> rules, List<Atom> facts) {
+		super(rules, facts, null);
+		this.analyzeFacts(facts);
+		this.analyzeRules(rules);
```

Comment:
I think we're embarking on a code-style discussion here ;-)
This is basically a matter of taste - Personally, I developed the habit of using fully qualified references as much as possible, and that includes qualifying member access using `this.`. In my opinion, it makes reading code a lot easier, especially if you're not using an IDE (which is a quite common case when we think about doing code reviews using github's online view, etc).
Since I know that's foremost the way I habitually do it, and not necessarily a universal best practice, I did some research into how it's perceived in the wider Java-Community.
I did find [one style guide](http://yohanan.org/steve/projects/java-code-conventions#SECTION000112000000000000000) that agrees with me:

> It is required to use the this object reference when accessing instance variables or calling non-static methods. Java assumes this in contexts where it is omitted; however, its explicit use removes any ambiguity. In addition, since Java is an object-oriented language it is considered good programming style to have all references (excluding local variables) to be prefaced by an object.

On the other hand, that style guide is also just someone's opinionated way of doing things, not a "universally accepted truth".
I guess [this thread](https://softwareengineering.stackexchange.com/questions/113430/what-is-the-accepted-style-for-using-the-this-keyword-in-java) illustrates that well - there are answers favoring both positions (consistently using `this` and not doing it). While the "accepted" answer is one that argues for omitting `this` since it's redundant, I'd also like to point out that the answer in question uses IDE features as it's main argument. As stated above, I personally think the argument `My IDE can disambiguate it anyway` is not valid since we're not using IDEs all the time (and shouldn't force people reading our code to do so).

So I guess ideally we should come to some "style-guide-like" decision here and then also incorporate that into the checkstyle config.

Reply:
Indeed, this is a matter of style (and taste) and there are reasons for both perspectives. So far, we mostly did omit `this.` if it is not needed. I also think that the `my IDE shows the difference` is a valid argument, because this is the way we usually look at our code.

Checkstyle does have a check that requires usage of `this.` (see https://checkstyle.org/config_coding.html#RequireThis ), but it does not have a check for the converse. Also note that in its default configuration it only warns if `this` is missing on overlapping variables, like in a constructor where you have `this.foo = foo` it warns if the `this.` is missing. One can configure it to warn every time, but even the checkstyle doc states "Warning: the Check is very controversial if 'validateOnlyOverlapping' option is set to 'false' and not that actual nowadays."

For Alpha, I therefore suggest to manually keep avoiding unnecessary `this`.